share/doc/revng/references/abi-definition.md

```diff
@@ -0,0 +1,213 @@
+
+## `ArgumentsCanBeSplitBetweenRegistersAndStack` (`bool`)
+
+States whether the ABI allows a single object that wouldn't fit into a single GPR (9+ bytes on 32-bit systems and 17+ bytes on 64-bit ones) to be partially passed in registers with the remainder placed on the stack if there are not enough registers to fit the entirety of it.
+
+As an example, let's say that there is a big object of type `Big` such that `sizeof(Big)` is equal to 16 bytes. On 32-bit system it would mean having to use four GPRs (`16 == 4 * 4`) to fit it.
+
+Let's look at an ABI that allocates four registers for passing function arguments (`r0-r3`). Then, for a function like
+```cpp
+void function(uint32_t, Big);
+```
+the `uint32_t` argument would be passed in the first GPR (`r0`).
+
+But that would also mean that the remaining three available GPRs are not enough to fit the entirety of the `Big` object, meaning it needs to either be split between the registers and the memory, or passed using the stack. That's exactly what this option states.
+
+## `BigArgumentsUsePointersToCopy` (`bool`)
+
+This allows a non-position-based ABI to take advantage of one of the main traits of position-based ABI: ability to use pointers to copy to pass big aggregates.
+
+This is one of the defining features of arm64 ABI, which allows it to get the best of both approaches - an ability to pass aggregates by copy in registers from the non-position-based approach while also being able to sometimes avoid a copy of a bigger aggregate from the position-based approach.
+
+## `NoRegisterArgumentsCanComeAfterStackOnes` (`bool`)
+
+States whether ABI allows a stack argument (mainly if it's too big to be placed in the registers) to precede other register arguments.
+
+For example, if there is a `Big` struct that has to use the stack, and
+a function like
+```cpp
+void function(Big, uint32_t);
+```
+if this value is set to true, both argument will be passed on stack, otherwise, only the struct will.
+
+## `AllowUnnaturallyAlignedTypesInRegisters` (`bool`)
+
+States whether ABI allows unnaturally aligned types[^1] to be passed in registers. When set to false, all the unnatural types will only be put on stack.
+
+## `CalleeIsResponsibleForStackCleanup` (`bool`)
+
+Specifies who is responsible for cleaning the stack after the function call. If equal to `true`, it's the callee, otherwise it the caller.
+
+## `FloatsUseGPRs` (`bool`)
+
+Setting this to true disables usage of vector registers entirely.
+
+## `StackArgumentsUseRegularStructAlignmentRules` (`bool`)
+
+Setting this to true allows stack arguments to use slots smaller than a word (makes them behave like a single struct instead of a bunch of arguments): so they don't get extended unnecessarily. A notable ABI using this is apple flavor of the aarch64.
```

Comment:
The concept of "slot" has not been defined, but seems rather important.

Reply:
Actually, I yoinked that terminology from the arm doc and have no idea how to even express it otherwise.

Looking at their doc, it doesn't look like they explain it anywhere either. They just use it, like:

> When an argument is assigned to a stack slot any unused padding bytes have unspecified value.

or

> `ldr x2, [x0, #__stack] // get stack slot pointer`

Reply:
I would then add a preamble such as "Certain ABIs (for example XXX) typically extend each stack argument to a slot (i.e., to a size of a GPR). This option...".

AFAIU this is an option that it specific  to Apple flavor of ARM's ABI and has no real general way to be described.
I think pretending this is not the case becomes counterproductive. I think the docs should be explicit and just say "this is an ad-hoc thing for...". Then, if in the future we can subsume this option in something more general we can get rid of that part, but as of now, pretending this is not ad-hoc, just makes the docs very mysterious.

Reply:
```
Most ABIs extend each stack argument to a GPR-sized slot. Setting this option to true, allows them occupy less space by behaving more like fields in a regular struct would.

This option is here exclusively for the apple flavor of the aarch64, since nothing else we support does anything similar.
```
?

Reply:
Works.

Reply:
Done.